### ISSUE TITLE
CFG-1 — Full training config schema with cross-section validation

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1574]
**Depends on:** CORE-1 (#1555), CORE-2 (#1556) · **Blocks:** TRN-1 (#1575), CLI-1 (#1579), FT-2 (#1581), FT-4 (#1637), TRN-5 (#1639)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy training has no config object. `mace/tools/arg_parser.py` is **1,417 lines** of untyped flags (`build_default_arg_parser()` exposes **184 dests** on `mace_run_train`, `build_preprocess_arg_parser()` **26**, **210 counted per parser** and 188 distinct since 22 are shared; 220 option strings carry them, the surplus being the ten `--swa_*`/`--stage_two_*` alias pairs), and the argparse `Namespace` doubles as the data model: `run()` in `mace/cli/run_train.py` reads **74 distinct `args.*` attributes** and assigns to `args` in place at **43 sites** — `args.heads = prepare_default_head(args)` (`:276`), `args.r_max = model_foundation.r_max.item()` (`:212`), `args.loss = "universal"` (`:435`), `args.avg_num_neighbors = get_avg_num_neighbors(...)` (`:835`) — and foundation-model presets overwrite user inputs behind an escape hatch (`args.lr = 0.0001`, `args.ema = True`, `args.ema_decay = 0.99999` at `:247-249`, skipped only when `--force_mh_ft_lr` is set, `arg_parser.py:597`). The YAML config path is **no longer optional or untested**: `configargparse` is a hard install dependency (`setup.cfg:32`), `--config` is a real flag (`arg_parser.py:22-27`, behind a `try/except ImportError` fallback that can no longer fire in a supported install), and workflow tests drive it (`tests/workflows/test_embedding_train.py:96-101`, `test_run_train.py`, `test_multifiles.py`). What is missing is not a file format — it is a **typed, validated, immutable object**. CORE-2 (#1556) built the machinery (pydantic-settings base class; TOML/YAML/JSON by extension; dotted CLI overrides; precedence defaults < file < CLI; unknown keys hard-error naming the key and its nearest valid neighbor; `to_resolved_dict()` with the file→resolved→file→resolved fixed-point property; the `ModelMetadata` schema). This ticket builds the **full training schema** on that machinery: one immutable, validated `ResolvedConfig` with `data / model / loss / optimizer / schedule / runtime` sections — the object at the head of the staged pipeline (`ResolvedConfig → DataStage → ModelStage → TrainStage`, TRN-1 (#1575)) — plus the cross-section validators and the `from_namespace(args)` shim that maps the pinned legacy flag surface onto it so the legacy CLI keeps working during coexistence. The resolved config (all defaults filled) is saved into model metadata for **every** model, not just foundation models.

**Interface & constraints:**

1. **Sections** (Pydantic models in `mace_core.config`; no torch/jax imports anywhere in the package). The group counts below size each section; they do **not** define its membership. The binding enumeration is P0-0's per-dest disposition table, which carries one row per argparse dest and fails the build when any dest lacks a `KEEP`/`MERGE`/`DROP` — naming a group here covers nothing on its own, which is how a whole optimizer's tuning surface went unmentioned:

| Section | Home | Owns (legacy flag group it replaces) |
|---|---|---|
| `data` | `config/data.py` — `DataConfig` + per-head `HeadDataConfig` (incl. `e0s: E0Spec`, below) | train/valid/test files (`--test_file`, `:472`, plus `--test_dir`), `valid_fraction`, heads, cutoff, `num_workers`, and the precomputed-statistics path `--statistics_file` (`:524`), which MERGEs into the prepared-dataset manifest — statistics travel with the shards rather than as a separate run-time path, and a legacy `statistics.json` is still readable through the compatibility parser (DATA-1 (#1569)/DATA-2 (#1570)) (the KEEP rows of the data/split flags; the **13 property-key flags** — one per `DefaultKeys` member, `mace/tools/default_keys.py:7-19` — follow CORE-3 (#1557)'s convention, not per-run flags) |
| `model` | `config/model.py` — `ModelConfig` (replaces the **67 distinct `args.*`** consumed by `mace/tools/model_script_utils.py`, i.e. `configure_model` and its helpers) | the architecture flags: `r_max`, `num_interactions`, `hidden_irreps` (+ `num_channels`/`max_L` shortcuts), `MLP_irreps` (`:264`, the hidden irreps of the last readout's MLP — a head-shape field, so it moves under the readout/head spec rather than staying a top-level knob), the non-linear-first-interaction-block option, …; the declared observable list (CORE-1 (#1555) names — declaring `energy, forces, dipole` is what later creates heads, ARCH-3 (#1562)). The legacy `--model` string enumerates ten names (`arg_parser.py:137-148`: `BOTNet`, `MACE`, `ScaleShiftMACE`, `PolarMACE`, `MACELES`, `ScaleShiftBOTNet`, `AtomicDipolesMACE`, `AtomicDielectricMACE`, `EnergyDipolesMACE`, `MagneticScaleShiftMACE`) — but only eight of them are classes: `BOTNet` (`:138`) and `ScaleShiftBOTNet` (`:143`) are accepted by argparse and then die in `configure_model` on `RuntimeError("… is deprecated, use MACE instead")` (`mace/tools/model_script_utils.py:377-381`). **The v1 model enum does not carry either**, so those two strings fail as unknown values through the normal config-validation path — a validation error naming the registered model names, raised before anything is built — instead of through a bespoke runtime raise inside model construction. In v1 the field is a registry name plus declared observables, never a switch that rewrites the observable set. `clebsch_gordan_basis` is **explicit model state** serialized with the checkpoint — one value for every device and backend, never inferred from installed packages |
| `loss` | `config/loss.py` — `LossConfig` | the **16 loss flags** (`--loss`, `--huber_delta`, and the seven base/stage-two weight pairs for energy, forces, magforces, virials, stress, dipole, polarizability) become **per-observable weights** keyed by CORE-1 (#1555) observable names (defaults from `ObservableSpec.default_loss_weight`) **plus a typed `params` field for a loss's own hyperparameters** — preserving the legacy `--huber_delta` (and any future per-loss knob) as config, not a flag. Tuning an existing loss (a weight or a hyperparameter) is a config change, never new code; loss *construction* is TRN-2 (#1576) — only the schema lives here |
| `optimizer` | `config/training.py` — `OptimizerConfig` | the **23 optimizer/scheduler/training-control flags** (`--optimizer` ∈ {`adam`, `adamw`, `schedulefree`}, `--beta`, batch sizes — both `--batch_size` and `--valid_batch_size` (`:926`), which are separate dests and separate fields — `--lr`, `--swa_lr`, `--weight_decay`, `--lr_params_factors`, `--freeze`, `--amsgrad`, `--scheduler`, `--lr_factor`, `--scheduler_patience`, `--lr_scheduler_gamma`, `--swa`, `--start_swa`, `--lbfgs`, `--ema`, `--ema_decay`, `--max_num_epochs`, `--patience`, `--eval_interval`, `--clip_grad`) plus the **10 `--swa_*`/`--stage_two_*` alias pairs**, which in develop are *option-string aliases of a single `dest`* (e.g. `arg_parser.py:809-815`), not two settings. Like `loss`, this section carries a **typed `params` field for the chosen optimizer's own hyperparameters**, and the schedulefree trio is its first occupant: `--beta1_schedulefree` (`:907`), `--beta2_schedulefree` (`:913`), `--warmup_steps_schedulefree` (`:919`). Naming `schedulefree` in the `--optimizer` enum does not carry its tuning surface — those three are fields, and each is a row in the per-dest table |
| `schedule` | `config/training.py` — `SchedulerConfig`, `StageTwoConfig` (ex-SWA), `EMAConfig` | LR schedule, stage-two switch, EMA; shaped for per-stage overrides (TRN-2 (#1576) executes them). The `--lbfgs` regime is a *stage kind*, not a boolean bolted onto the optimizer — TRN-5 (#1639) owns its execution, this ticket owns the field |
| `runtime` | `config/runtime.py` — runtime/output | the **23 run/infra flags**: `name`, `seed`, `device`, `default_dtype`, `log_level`, `distributed`, `launcher`, `num_workers`, `pin_memory`, the checkpoint-retention trio, `plot`/`plot_frequency`, …; `work_dir`/`log_dir`/`model_dir`/`checkpoints_dir`/`results_dir`/`downloads_dir` MERGE into a **single work-dir layout convention** (document it). The three backend-selection dests — `--enable_cueq`, `--enable_oeq` and `--only_cueq` (`:1089`) — are **not** runtime fields: they MERGE into BKD-1 (#1560)'s single backend-name field, resolved once at model build time. `--only_cueq` loses its meaning entirely in that move, since it exists only to choose between *building* in cueq layout (`mace/tools/model_script_utils.py:180`) and *converting* after build (the conversion it gates at `mace/cli/run_train.py:881,1175`), and canonical weights remove the conversion. Both facts are dest rows in the P0-0 table, owned by BKD-1 (#1560) |

`ResolvedConfig` composes the six sections and reserves a `finetune` section whose full field set lands with FT-1 (#1580)/FT-2 (#1581)/FT-4 (#1637) — but the validators assigned to this ticket (E0s and pseudolabel rules below) ship here.

2. **`E0Spec` — the typed union replacing the stringly-typed `--E0s` paths** (in `mace_core.config`, alongside `DataConfig`; E0 values are IEEE fp64 end to end):

```python
MissingElementPolicy = Literal["error", "average", "zero"]   # default "error" everywhere

class E0sTable(BaseModel):            # explicit values; also the parsed form of a JSON file
    kind: Literal["table"] = "table"
    values: dict[int, float]          # z -> E0, eV

class E0sIsolatedAtoms(BaseModel):    # from config_type=IsolatedAtom entries in the head's training file
    kind: Literal["isolated_atoms"] = "isolated_atoms"
    on_missing_energy: Literal["error", "zero"] = "error"    # legacy warns + 0.0 (mace/data/utils.py:333-337)

class E0sAverage(BaseModel):          # least squares over the head's training energies
    kind: Literal["average"] = "average"
    # legacy logs an error and zeroes every element on LinAlgError
    # (mace/data/utils.py:381-387); v1 hard-errors

class E0sFromFoundation(BaseModel):   # copy from FM artifact metadata — never by buffer reflection
    kind: Literal["foundation"] = "foundation"
    head: str | None = None           # None -> error if FM is multi-head (legacy takes head 0 and
                                      # only logs it, mace/cli/run_train.py:510-514, :528-532, :560-565)
    missing: MissingElementPolicy = "error"   # elements not covered by the FM

class E0sEstimated(BaseModel):        # FM-corrected least squares
    kind: Literal["estimated"] = "estimated"
    head: str | None = None
    missing: MissingElementPolicy = "error"

E0Spec = Annotated[Union[E0sTable, E0sIsolatedAtoms, E0sAverage,
                         E0sFromFoundation, E0sEstimated], Field(discriminator="kind")]
```

Per-head default: `HeadDataConfig.e0s = E0sIsolatedAtoms()` (matches today's implicit first choice). The E0 *resolver* and its provenance object are consumed by DataStage/ARCH-3 (#1562), not built here — this ticket owns the spec and its validators.

3. **Cross-section validators** (the point of the ticket — every one replaces a legacy fallback that keeps training with a loud error):
   - `e0s.kind in {"foundation", "estimated"}` **requires** a foundation-model section to be configured (legacy asserts this only for `estimated`, `mace/cli/run_train.py:527`).
   - Dipole/dielectric model configs **forbid** an `e0s` field entirely. Develop no longer rewrites it silently: under `--finetune_dipoles_polarizabilities` it warns and downgrades `foundation`/`estimated` to `average` (`mace/cli/run_train.py:153-159`) because `AtomicDielectricMACE` has no `atomic_energies_fn`. v1 makes the same situation a **validation error at config time** rather than a warning at run time.
   - The implicit replay head is fixed to `E0sFromFoundation(missing="error")`.
   - Pseudolabel rule: `labels_from` (explicit reuse of a prior label artifact, `Path`) is **mutually exclusive** with `enabled=True` (in-run generation) — one validator, fields per the finetune-section spec that FT-2 (#1581) implements; pseudolabel provenance serializes with the resolved-config metadata.
   - LBFGS rule: an LBFGS stage forbids `ema` and forbids a `ReduceLROnPlateau`-style scheduler on that stage — legacy simply replaces the optimizer object after the scheduler and EMA were already built (`mace/cli/run_train.py:971-976`), leaving both attached to a full-batch closure that steps once per epoch. TRN-5 (#1639) executes the regime; the validator lives here.
   - General principle: Pydantic checks compatibility **between** sections, but the training section stays unaware of whether the run is fine-tuning — minimal objects cross section boundaries.

4. **`from_namespace(args)` shim.** Maps the full pinned legacy flag surface onto `ResolvedConfig` — the complete mapping is this ticket's deliverable, and it is **per dest, not per group**: all 184 `mace_run_train` dests, each landing on a field, on a named MERGE target, or on a recorded DROP, driven by P0-0's mechanical KEEP/MERGE/DROP inventory (do not re-derive dispositions; implement them). Since an argparse `Namespace` is itself keyed on dest, the shim's domain and the inventory's key are the same object, and the test in the acceptance list is a set comparison against it. The 10 `--swa_*`/`--stage_two_*` alias pairs are one dest apiece and collapse to one field each, keeping the `stage_two` spelling. Foundation presets become **declared defaults** (visible in the resolved config, overridable by explicit user settings) — never post-parse mutations, and there is no `--force_mh_ft_lr` equivalent because there is nothing to force. This shim is what keeps the legacy CLI working during coexistence and what CLI-1 (#1579)'s v0.3-args→v1-config migration helper builds on.

5. **Invariants.** `ResolvedConfig` is immutable after validation; unknown keys are hard errors; all defaults are filled in the resolved form; CORE-2 (#1556)'s fixed-point property (file→resolved→file→resolved) must hold on the full training schema; the resolved config serializes into `ModelMetadata` for every model. Sections must be consumable piecemeal: `mace data prepare` (CLI-1 (#1579)/DATA) reads the same `data`/`model` sections instead of duplicating the 26 preprocess dests, so no section may assume the full training context exists.

**Recorded DROPs (this ticket's rows in the P0-0 matrix):** `--save_cpu` (`arg_parser.py:1085`; v1 checkpoints are device-agnostic, TRN-4 (#1578)); `--force_mh_ft_lr` (`:604`; presets are declared defaults, nothing to force); `--plot_interaction_e` (`:107`; a legacy plotting mode with no v1 consumer, CLI-1 (#1579)); the `--swa*` option-string spellings (the `--stage_two*` spelling is the KEEP); the `--model` values `BOTNet` (`:138`) and `ScaleShiftBOTNet` (`:143`), which name no class in develop and only reach a deprecation `RuntimeError` (`mace/tools/model_script_utils.py:377-381`).

**Task:**
1. `mace_core.config`: the six section models (`data.py`, `model.py`, `loss.py`, `training.py`, `runtime.py`) + `ResolvedConfig` composing them on CORE-2 (#1556)'s `ReforgeBaseConfig` machinery (three formats, dotted overrides, precedence, unknown-key errors).
2. `E0Spec` discriminated union + `HeadDataConfig.e0s` with the defaults above.
3. Cross-section validators: foundation/estimated⇒FM-section, dipole-forbids-e0s, replay-head pin, `labels_from` ⊕ `enabled`, LBFGS ⊕ EMA/plateau-scheduler, with error messages naming the offending fields.
4. `from_namespace(args)` covering the pinned legacy flag surface (including the 10 alias-pair collapses), presets-as-declared-defaults, and the single work-dir layout convention replacing the six `*_dir` flags.
5. Resolved-config→`ModelMetadata` wiring (CORE-2 (#1556) schema) + round-trip and fixed-point tests over the full schema.

**Out of scope:** the config *machinery* (done — CORE-2 (#1556)); the stages that consume `ResolvedConfig` and the training loop (TRN-1 (#1575)); loss construction and per-stage schedule execution (TRN-2 (#1576)); the LBFGS regime's execution (TRN-5 (#1639)); the E0 resolver and provenance implementation (DataStage/ARCH-3 (#1562) consume the spec); the full `finetune` field set, pseudolabel generation and the MDP fine-tune mode (FT-2 (#1581)/FT-4 (#1637)); CLI entry points and the migration helper (CLI-1 (#1579)); backend/precision config (BKD-1 (#1560)/BKD-2 (#1566)).

**Acceptance criteria:**
- [ ] `ResolvedConfig` loads from TOML, YAML and JSON; dotted CLI overrides reach every section; precedence defaults < file < CLI holds; an unknown key hard-errors naming the key and its nearest valid neighbor.
- [ ] file→resolved→file→resolved is a fixed point on the full training schema (property test).
- [ ] All five `E0Spec` kinds parse via the discriminator; `foundation`/`estimated` without a foundation-model section is a validation error; a dipole/dielectric model config with `e0s` set is a validation error; `E0sAverage` documents hard-error (not the legacy all-zero fallback) on singular systems.
- [ ] `labels_from` set together with `enabled=True` is a validation error; an LBFGS stage with EMA or a plateau scheduler is a validation error.
- [ ] `from_namespace(args)` round-trips the full pinned legacy flag surface: the test re-derives the **184 `mace_run_train` dests** from `build_default_arg_parser()` and asserts each is mapped, MERGEd or a recorded DROP — a set comparison, so a dest added to `arg_parser.py` fails the test rather than passing unnoticed. The 10 alias pairs resolve to one field each; foundation presets appear as overridable declared defaults, and no code path mutates a config after validation.
- [ ] The resolved config (all defaults filled) serializes into `ModelMetadata` and back.
- [ ] `mace_core.config` imports neither torch nor jax (asserted).

**Verify:**
```bash
python -m pytest packages/mace-core/tests/test_training_config.py -v          # sections, validators, fixed point
python -m pytest packages/mace-core/tests/test_e0_spec.py -v                  # union + cross-section rules
python -m pytest packages/mace-core/tests/test_config_from_namespace.py -v    # legacy flag surface round-trip
```
**Review focus:** the cross-section validators — each one exists to replace a legacy keep-going fallback (E0 zeroing, the dielectric E0s downgrade, the FM head-0 pick, preset overwrites) with a loud error, so review them against the cited legacy lines; and completeness of `from_namespace` against the pinned flag inventory (a missed flag silently changes training behavior under `--engine legacy`-to-v1 migration).

---
